File: 2347-count-nodes-equal-to-average-of-subtree/count-nodes-equal-to-average-of-subtree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    int result=0;
    pair<int,int> solve(TreeNode* root,int sum,int count)
    {
        if(!root)
        return {0,0};

        pair<int,int> Left=solve(root->left,sum,count);
        pair<int,int> Right=solve(root->right,sum,count);

        sum=root->val+Left.second+Right.second;
        count=1+Left.first+Right.first;

        if(sum/count==root->val)
        result++;

        return {count,sum};
    }

    int averageOfSubtree(TreeNode* root) 
    {
        solve(root,0,0);
        return result;
    }
};

File: 2347-count-nodes-equal-to-average-of-subtree/count-nodes-equal-to-average-of-subtree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int count;
    private int[] solve(TreeNode root) {
        if(root == null) {
            return new int[]{0, 0};
        }
        
        int[] leftResult = solve(root.left);
        int[] rightResult = solve(root.right);
        
        int subtreeSum = root.val +  leftResult[0] + rightResult[0];
        int subtreeSize = 1 + leftResult[1] + rightResult[1];
        
        if(root.val == subtreeSum / subtreeSize) {
            count++;
        }
        
        return new int[]{subtreeSum, subtreeSize};
    }
    
    public int averageOfSubtree(TreeNode root) {
        count = 0;
        solve(root);
        return count;
    }
}